fearless/Common/QRCoding/QRCreationOperation.swift

```diff
@@ -0,0 +1,70 @@
+import Foundation
+import CoreImage
+import RobinHood
+
+enum QRCreationOperationError: Error {
+    case generatorUnavailable
+    case generatedImageInvalid
+    case bitmapImageCreationFailed
+}
+
+public final class QRCreationOperation: BaseOperation<UIImage> {
+    let payload: Data
+    let qrSize: CGSize
+
+    init(payload: Data, qrSize: CGSize) {
+        self.payload = payload
+        self.qrSize = qrSize
+
+        super.init()
+    }
+
+    override public func main() {
+        super.main()
+
+        guard let filter = CIFilter(name: "CIQRCodeGenerator") else {
+            if !isCancelled {
+                result = .failure(QRCreationOperationError.generatorUnavailable)
+            }
+
+            return
+        }
+
+        filter.setValue(payload, forKey: "inputMessage")
+        filter.setValue("M", forKey: "inputCorrectionLevel")
+
+        guard let qrImage = filter.outputImage else {
+            if !isCancelled {
+                result = .failure(QRCreationOperationError.generatedImageInvalid)
+            }
+
+            return
+        }
+
+        let transformedImage: CIImage
+
+        if qrImage.extent.size.width * qrImage.extent.height > 0.0 {
+            let transform = CGAffineTransform(
```

Comment:
didn't you miss original image ratio there?
I think you need to calculate one scale for both x and y axis

Reply:
It's okay. This logic is already checked.